lib/_stream_readable.js

```diff
@@ -1070,13 +1072,84 @@ Readable.prototype.wrap = function(stream) {
 };
 
 Readable.prototype[SymbolAsyncIterator] = function() {
-  if (createReadableStreamAsyncIterator === undefined) {
-    createReadableStreamAsyncIterator =
-      require('internal/streams/async_iterator');
+  let stream = this;
+
+  if (typeof stream.read !== 'function') {
+    // v1 stream
+    const src = stream;
+    stream = new Readable({
+      objectMode: true,
+      destroy(err, callback) {
+        destroyImpl.destroyer(src, err);
+        callback();
+      }
+    }).wrap(src);
   }
-  return createReadableStreamAsyncIterator(this);
+
+  const iter = createAsyncIterator(stream);
+  iter.stream = stream;
+  return iter;
 };
 
+async function* createAsyncIterator(stream) {
+  const state = stream._readableState;
+
+  if ((state && state.endEmitted) ||
+      (state && state.closeEmitted)) {
+    // TODO (ronag): What if errorEmitted?
+    // TODO (ronag): Wait for closeEmitted?
+    // TODO (ronag): if (closeEmitted && !endEmitted) ERR_PREMATURE_CLOSE?
+    stream.on('error', nop);
+    return;
+  }
+
+  let callback = nop;
+  let error;
+
+  stream
+    .on('readable', () => {
+      callback();
+      callback = nop;
+    })
+    .on('error', (err) => {
+      error = err;
+      callback();
+      callback = nop;
+    })
+    .on('end', () => {
+      error = error || null;
+      callback();
+      callback = nop;
+    })
+    .on('close', () => {
+      error = error || null;
+      // TODO (ronag): ERR_PREMATURE_CLOSE?
+      callback();
+      callback = nop;
+    });
+
+  try {
+    while (error === undefined) {
+      const chunk = stream.read();
+      if (chunk !== null) {
+        yield chunk;
+      } else {
+        await new Promise((resolve) => {
+          callback = resolve;
+        });
+      }
+    }
+    // TODO (ronag): Wait for closeEmitted?
+  } catch (err) {
+    error = err;
+  } finally {
+    destroyImpl.destroyer(stream, error);
```

Comment:
Random question, what if the stream's _destroy throws in terms of what error the async iterator consumer sees?

Reply:
I think throwing inside `_destroy` is currently pretty much undefined behavior in general.

Reply:
I think the async iterator consumer will see that error.

Reply:
So the initial error is swallowed and the error thrown from `destroy` gets emitted?

Reply:
that would be my guess

Reply:
I am not a fan of this behavior but it's not new.

Reply:
I mean... we shouldn't be throwing in `_destroy`. I guess we could in theory add a try/catch inside the `destroy` implementation (which calls `_destroy`). I'm just not sure what to do with it? Unhandled exception?

Reply:
I think it's an unrelated breaking change so I wouldn't put it here anyway, but yeah - probably